File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.cpp
class Solution {
public:
    bool ok(long long t, vector<int>& time, int total){
        long long count = 0;
        for(int& i : time) count += (t / i);
        return count >= total;
    }
    long long minimumTime(vector<int>& time, int totalTrips) {
        long long l = 0;
        long long min = *min_element(time.begin(), time.end());
        long long r = min * totalTrips;
        while(l < r){
            long long mid = (l + r)/2;
            if(ok(mid,time,totalTrips)) r = mid;
            else l = mid + 1;
        }
        return l;
    }
};

File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.py
class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        def ok(t : int, time : List[int], total : int) : 
            count = 0
            for i in time : 
                count += (t // i)
            return count >= total 
        
        l = 0
        mini = min(time)
        r = mini * totalTrips
        while l < r : 
            mid = (l + r) >> 1
            if ok(mid, time, totalTrips) : 
                r = mid
            else : 
                l = mid + 1
            
        return l